crates/flowctl/src/source/unbundle.rs

```diff
@@ -0,0 +1,338 @@
+            let mut v2 = Vec::new();
+            let mut deser = serde_json::Deserializer::from_slice(&v);
+            serde_transcode::transcode(&mut deser, &mut serde_yaml::Serializer::new(&mut v2))
+                .context("serializing YAML")?;
+
+            Ok(v2)
+        }
+    }
+}
+
+async fn try_write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    // Is the file allowed to exist before we create it?
+    let create_new = match existing {
+        Existing::Abort | Existing::Keep => true,
+        _ => false,
+    };
+    let result = fs::OpenOptions::new()
+        .create_new(create_new)
+        .create(true)
+        .write(true)
+        .open(path)
+        .await;
+    match result {
+        Ok(mut file) => {
+            file.write_all(contents).await?;
+            Ok(())
+        }
+        Err(err)
+            if err.kind() == std::io::ErrorKind::AlreadyExists && existing == Existing::Keep =>
+        {
+            tracing::info!(path = %path.display(), "skipping writing file because it already exists and --existing=skip");
+            Ok(())
+        }
+        // This branch also handles the case of `Existing::Error` if the file already exists.
+        Err(err) => Err(anyhow::anyhow!("unable to open file: {}", err)),
+    }
+}
+
+async fn write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    match try_write_file(path, contents, existing).await {
+        Ok(()) => {
+            tracing::info!(path = %path.display(), "wrote file");
+            Ok(())
+        }
+        Err(err) => {
+            tracing::error!(path = %path.display(), error = %err, "failed to write file");
+            Err(err).context(format!("failed to write file: '{}'", path.display()))
+        }
+    }
+}
+
+async fn parse_catalog_spec(flow_yaml_path: &Path) -> Result<models::Catalog, anyhow::Error> {
+    let mut existing = fs::File::open(flow_yaml_path).await?;
+    let mut contents = Vec::with_capacity(16 * 1024);
+    existing.read_to_end(&mut contents).await?;
+    let cat = sources::parse_catalog_spec(&contents)?;
+    Ok(cat)
+}
+
+/// Returns the last path component for a given catalog name, which must be a valid catalog name.
+fn base_name(catalog_name: &impl AsRef<str>) -> &str {
+    catalog_name
+        .as_ref()
+        .rsplit_once("/")
+        .expect("catalog names have at least one '/'")
+        .1
+}
+
+async fn indirect_endpoint_config(
+    config: &mut models::ConnectorConfig,
+    dir: &Path,
+    filename: String,
+    serializer: &Serializer,
+    existing: Existing,
+) -> anyhow::Result<()> {
+    let file_path = dir.join(&filename);
+    let contents = serializer.to_vec(&config.config)?;
+
+    write_file(&file_path, &contents, existing).await?;
+    config.config = RawValue::from_string(Value::String(filename).to_string()).unwrap();
+    Ok(())
+}
+
+async fn maybe_indirect_schema(
```

Comment:
I don't understand what this is about (i.e I can see what's happening, just not why), so maybe a comment would be useful here to explain why this is needed? Or maybe just an answer to this GH comment so I can understand :P

Reply:
A bundled catalog is one that has all schema resources inlined. The act of unbundling is taking those inlined schemas and writing them out to separate files, replacing the inlined schema with a relative URL pointing to the file. The `maybe_` is because, if the schema is very small, then we just keep it inlined. The overall goal of unbundling is to leave the user with a directory of Flow specs that's easy to navigate and read, so they can edit them and potentially re-publish them.

Reply:
Ah sorry, I meant the stuff about `$id` and `$defs`. The part about keeping very small schemas inlined is great though, that should make things much less fragmented

Reply:
Every JSON-schema that we bundle includes a top-level `$id` keyword (part of the JSON-schema spec) which identifies the canonical URI (CURI) from which we got the schema.

Every schema must have a CURI, so in cases where the schema was originally inlined we'll invent a CURI for it using a scheme I'll skip over in this comment.

That CURI often isn't meaningful to the user can can be removed, as it looks weird to leave it. BUT, if $defs are included in the JSON-schema, it may be a bundled representation of _multiple_ source files, and `$id` may play an important role as the base URL against which references are resolved, and it would be incorrect to remove it.